socketTimeoutUnits and establishConnectionTimeoutUnits in HttpClientBuilder are not used

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

When creating an http client, you can set socket and establish connection timeouts and time units, like this:
```
HttpClientBuilder httpClientBuilder = new HttpClientBuilder(); 
httpClientBuilder
    .socketTimeout(myConnectionSocketTimeout, TimeUnit.SECONDS)
    .establishConnectionTimeout(myConnectionEstablishConnectionTimeout, TimeUnit.SECONDS)
```
In the Keycloak HttpClientBuilder these time unit values are never used, because they are not supported by the underlying Apache implementation.
This is confusing to the end user.

This bug is present since Keycloak 1.4.0.Final (9 years ago)

### Version

24.0.2

### Regression

- [ ] The issue is a regression

### Expected behavior

The time units are used to calculate the amount of milliseconds.
```
HttpClientBuilder httpClientBuilder = new HttpClientBuilder(); 
httpClientBuilder
    .socketTimeout(2, TimeUnit.SECONDS)
    .establishConnectionTimeout(1, TimeUnit.SECONDS)
```
will result inside [HttpClientBuilder ](https://github.com/keycloak/keycloak/blame/6a020d93f15c5c03e9b76a5e02b51e5391aade6a/services/src/main/java/org/keycloak/connections/httpclient/HttpClientBuilder.java#L252) in
```
RequestConfig requestConfig = RequestConfig.custom()
    .setConnectTimeout(1000)
    .setSocketTimeout(2000);
```

### Actual behavior

The time units are ignored and the amount is always in milliseconds.
For example:
```
HttpClientBuilder httpClientBuilder = new HttpClientBuilder(); 
httpClientBuilder
    .socketTimeout(2, TimeUnit.SECONDS)
    .establishConnectionTimeout(1, TimeUnit.SECONDS)
```
will result inside [HttpClientBuilder](https://github.com/keycloak/keycloak/blame/6a020d93f15c5c03e9b76a5e02b51e5391aade6a/services/src/main/java/org/keycloak/connections/httpclient/HttpClientBuilder.java#L252) in
```
RequestConfig requestConfig = RequestConfig.custom()
    .setConnectTimeout(1)
    .setSocketTimeout(2);
```
This will fail very fast!

### How to Reproduce?

```
import org.apache.http.impl.client.CloseableHttpClient;
import org.keycloak.broker.provider.util.SimpleHttp;
import org.keycloak.connections.httpclient.HttpClientBuilder;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

class Scratch {
    public static void main(String[] args) throws IOException {
        HttpClientBuilder httpClientBuilder = new HttpClientBuilder();
        httpClientBuilder
                .socketTimeout(2, TimeUnit.SECONDS)
                .establishConnectionTimeout(1, TimeUnit.SECONDS);
        CloseableHttpClient httpClient = httpClientBuilder.build();
        try (SimpleHttp.Response response = SimpleHttp.doGet("https://httpbin.org/get", httpClient).asResponse()) {
            System.out.println(response.asString());
        }
    }
}
```

Will fail almost immediately with:

```
Exception in thread "main" org.apache.http.conn.ConnectTimeoutException: Connect to httpbin.org:443 [httpbin.org/52.200.250.127, httpbin.org/34.200.110.94, httpbin.org/54.242.212.250, httpbin.org/34.204.34.199, httpbin.org/52.203.147.106, httpbin.org/44.209.252.174, httpbin.org/3.234.161.234, httpbin.org/18.210.66.151] failed: Connect timed out
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:151)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:376)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:393)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
	at org.keycloak.broker.provider.util.SimpleHttp.makeRequest(SimpleHttp.java:299)
	at org.keycloak.broker.provider.util.SimpleHttp.asResponse(SimpleHttp.java:229)
	at Scratch.main(scratch_3.java:15)
Caused by: java.net.SocketTimeoutException: Connect timed out
	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:551)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:602)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
	at java.base/java.net.Socket.connect(Socket.java:633)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:368)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:142)
	... 13 more
```

### Anything else?

_No response_

--- a/services/src/main/java/org/keycloak/connections/httpclient/HttpClientBuilder.java
+++ b/services/src/main/java/org/keycloak/connections/httpclient/HttpClientBuilder.java
@@ -250,8 +250,8 @@ public class HttpClientBuilder {
             }
 
             RequestConfig requestConfig = RequestConfig.custom()
-                    .setConnectTimeout((int) establishConnectionTimeout)
-                    .setSocketTimeout((int) socketTimeout)
+                    .setConnectTimeout((int) TimeUnit.MILLISECONDS.convert(establishConnectionTimeout, establishConnectionTimeoutUnits))
+                    .setSocketTimeout((int) TimeUnit.MILLISECONDS.convert(socketTimeout, socketTimeoutUnits))
                     .setExpectContinueEnabled(expectContinueEnabled).build();
 
             org.apache.http.impl.client.HttpClientBuilder builder = HttpClients.custom()

--- a/services/src/test/java/org/keycloak/connections/httpclient/HttpClientBuilderTest.java
+++ b/services/src/test/java/org/keycloak/connections/httpclient/HttpClientBuilderTest.java
@@ -1,0 +1,56 @@
+package org.keycloak.connections.httpclient;
+
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.lang.reflect.Field;
+import java.util.concurrent.TimeUnit;
+
+public class HttpClientBuilderTest {
+
+    @Test
+    public void testDefaultBuilder() throws NoSuchFieldException, IllegalAccessException {
+        CloseableHttpClient httpClient = new HttpClientBuilder().build();
+
+        RequestConfig requestConfig = getRequestConfig(httpClient);
+
+        Assert.assertEquals("Default socket timeout is -1 and can be converted by TimeUnit", -1, requestConfig.getSocketTimeout());
+        Assert.assertEquals("Default connect timeout is -1 and can be converted by TimeUnit", -1, requestConfig.getConnectTimeout());
+    }
+
+    @Test
+    public void testTimeUnitSeconds() throws NoSuchFieldException, IllegalAccessException {
+        HttpClientBuilder httpClientBuilder = new HttpClientBuilder();
+        httpClientBuilder
+                .socketTimeout(2, TimeUnit.SECONDS)
+                .establishConnectionTimeout(1, TimeUnit.SECONDS);
+        CloseableHttpClient httpClient = httpClientBuilder.build();
+
+        RequestConfig requestConfig = getRequestConfig(httpClient);
+
+        Assert.assertEquals("Socket timeout is converted to milliseconds", 2000, requestConfig.getSocketTimeout());
+        Assert.assertEquals("Connect timeout is converted to milliseconds", 1000, requestConfig.getConnectTimeout());
+    }
+
+    @Test
+    public void testTimeUnitMilliSeconds() throws NoSuchFieldException, IllegalAccessException {
+        HttpClientBuilder httpClientBuilder = new HttpClientBuilder();
+        httpClientBuilder
+                .socketTimeout(2000, TimeUnit.MILLISECONDS)
+                .establishConnectionTimeout(1000, TimeUnit.MILLISECONDS);
+        CloseableHttpClient httpClient = httpClientBuilder.build();
+
+        RequestConfig requestConfig = getRequestConfig(httpClient);
+
+        Assert.assertEquals("Socket timeout is still in milliseconds", 2000, requestConfig.getSocketTimeout());
+        Assert.assertEquals("Connect timeout is still in milliseconds", 1000, requestConfig.getConnectTimeout());
+    }
+
+    private static RequestConfig getRequestConfig(CloseableHttpClient httpClient) throws NoSuchFieldException, IllegalAccessException {
+        Field defaultConfig = httpClient.getClass().getDeclaredField("defaultConfig");
+        defaultConfig.setAccessible(true);
+        return (RequestConfig) defaultConfig.get(httpClient);
+    }
+}

